File: packages/types/src/acp/session.ts
/**
 * Session setup and lifecycle (`initialize`, `session/new|prompt|cancel|
 * resume`), todo #350. Mirrors `mainframe-types/src/acp/session.rs` — see
 * that file's module doc for which fields are kept opaque (`unknown`) and
 * why (auth, elicitation, MCP wiring, session config options, the resume
 * cursor scheme are orthogonal to this task's payload grammar).
 */
import { z } from 'zod';
import { ContentBlockSchema } from './content.js';

export type SessionId = string;
export type ProtocolVersion = number;

/** The ACP v2 protocol version this vendored subset targets (spec decision 2: frozen snapshot `d0370de50e16`). */
export const PINNED_PROTOCOL_VERSION: ProtocolVersion = 2;

export const ImplementationSchema = z
  .object({
    name: z.string(),
    title: z.string().optional(),
    version: z.string(),
  })
  .loose();
export type Implementation = z.infer<typeof ImplementationSchema>;

export const InitializeRequestSchema = z
  .object({
    protocolVersion: z.number().int().nonnegative(),
    info: ImplementationSchema,
    capabilities: z.unknown().optional(),
    _meta: z.record(z.string(), z.unknown()).optional(),
  })
  .loose();
export type InitializeRequest = z.infer<typeof InitializeRequestSchema>;

export const InitializeResponseSchema = z
  .object({
    protocolVersion: z.number().int().nonnegative(),
    info: ImplementationSchema,
    capabilities: z.unknown().optional(),
    authMethods: z.array(z.unknown()).optional(),
    _meta: z.record(z.string(), z.unknown()).optional(),
  })
  .loose();
export type InitializeResponse = z.infer<typeof InitializeResponseSchema>;

export const NewSessionRequestSchema = z
  .object({
    cwd: z.string(),
    additionalDirectories: z.array(z.string()).optional(),
    mcpServers: z.array(z.unknown()).optional(),
    _meta: z.record(z.string(), z.unknown()).optional(),
  })
  .loose();
export type NewSessionRequest = z.infer<typeof NewSessionRequestSchema>;

export const NewSessionResponseSchema = z
  .object({
    sessionId: z.string(),
    configOptions: z.array(z.unknown()).optional(),
    _meta: z.record(z.string(), z.unknown()).optional(),
  })
  .loose();
export type NewSessionResponse = z.infer<typeof NewSessionResponseSchema>;

export const PromptRequestSchema = z
  .object({
    sessionId: z.string(),
    prompt: z.array(ContentBlockSchema),
    _meta: z.record(z.string(), z.unknown()).optional(),
  })
  .loose();
export type PromptRequest = z.infer<typeof PromptRequestSchema>;

/**
 * Acceptance of a prompt, distinct from turn completion (spec: "`session/
 * prompt` acceptance is separate from turn completion"). `_meta` carries the
 * Mainframe queued-state extension when the prompt was accepted mid-turn.
 */
export const PromptResponseSchema = z
  .object({
    _meta: z.record(z.string(), z.unknown()).optional(),
  })
  .loose();
export type PromptResponse = z.infer<typeof PromptResponseSchema>;

export const CancelSessionNotificationSchema = z
  .object({
    sessionId: z.string(),
    _meta: z.record(z.string(), z.unknown()).optional(),
  })
  .loose();
export type CancelSessionNotification = z.infer<typeof CancelSessionNotificationSchema>;

export const ResumeSessionRequestSchema = z
  .object({
    sessionId: z.string(),
    cwd: z.string(),
    additionalDirectories: z.array(z.string()).optional(),
    mcpServers: z.array(z.unknown()).optional(),
    /** Opaque replay cursor — the cursor scheme itself is group E's concern (plan task 15). */
    replayFrom: z.unknown().optional(),
    _meta: z.record(z.string(), z.unknown()).optional(),
  })
  .loose();
export type ResumeSessionRequest = z.infer<typeof ResumeSessionRequestSchema>;

export const ResumeSessionResponseSchema = z
  .object({
    configOptions: z.array(z.unknown()).optional(),
    _meta: z.record(z.string(), z.unknown()).optional(),
  })
  .loose();
export type ResumeSessionResponse = z.infer<typeof ResumeSessionResponseSchema>;

File: packages/core-rs/crates/mainframe-types/src/acp/session.rs
//! Session setup and lifecycle methods (`initialize`, `session/new|prompt|
//! cancel|resume`), scoped to the chat facade's needs. `ClientCapabilities`/
//! `AgentCapabilities`/`mcpServers`/`additionalDirectories`/`configOptions`/
//! `replayFrom` are kept as opaque `serde_json::Value` — their substructure
//! (auth, elicitation, MCP server wiring, session config options, the resume
//! cursor scheme) is orthogonal to the payload grammar this task vendors and
//! is either declined by the spec (auth, fs/terminal) or owned by a later
//! group (group E's cursor scheme, task 15). The Mainframe extension
//! capabilities that *are* this task's concern live in `extensions.rs` and
//! ride under `_meta["_mainframe.dev"]`.

use serde::{Deserialize, Serialize};
use serde_json::Value;

pub type SessionId = String;
pub type ProtocolVersion = u16;

/// The ACP v2 protocol version this vendored subset targets (spec decision 2:
/// frozen snapshot `d0370de50e16`, `schema/v2/meta.json` `"version": 2`).
pub const PINNED_PROTOCOL_VERSION: ProtocolVersion = 2;

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Implementation {
    pub name: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub title: Option<String>,
    pub version: String,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct InitializeRequest {
    pub protocol_version: ProtocolVersion,
    pub info: Implementation,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub capabilities: Option<Value>,
    #[serde(default, skip_serializing_if = "Option::is_none", rename = "_meta")]
    pub meta: Option<Value>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct InitializeResponse {
    pub protocol_version: ProtocolVersion,
    pub info: Implementation,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub capabilities: Option<Value>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub auth_methods: Option<Vec<Value>>,
    #[serde(default, skip_serializing_if = "Option::is_none", rename = "_meta")]
    pub meta: Option<Value>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct NewSessionRequest {
    pub cwd: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub additional_directories: Option<Vec<String>>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub mcp_servers: Option<Vec<Value>>,
    #[serde(default, skip_serializing_if = "Option::is_none", rename = "_meta")]
    pub meta: Option<Value>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct NewSessionResponse {
    pub session_id: SessionId,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub config_options: Option<Vec<Value>>,
    #[serde(default, skip_serializing_if = "Option::is_none", rename = "_meta")]
    pub meta: Option<Value>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PromptRequest {
    pub session_id: SessionId,
    pub prompt: Vec<super::content::ContentBlock>,
    #[serde(default, skip_serializing_if = "Option::is_none", rename = "_meta")]
    pub meta: Option<Value>,
}

/// Acceptance of a prompt, distinct from turn completion (spec: "`session/
/// prompt` acceptance is separate from turn completion"). `meta` carries the
/// Mainframe queued-state extension (`extensions::QueuedPromptState`) when the
/// prompt was accepted mid-turn.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct PromptResponse {
    #[serde(default, skip_serializing_if = "Option::is_none", rename = "_meta")]
    pub meta: Option<Value>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CancelSessionNotification {
    pub session_id: SessionId,
    #[serde(default, skip_serializing_if = "Option::is_none", rename = "_meta")]
    pub meta: Option<Value>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ResumeSessionRequest {
    pub session_id: SessionId,
    pub cwd: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub additional_directories: Option<Vec<String>>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub mcp_servers: Option<Vec<Value>>,
    /// Opaque replay cursor (`ReplayFrom`) — the cursor scheme itself is
    /// group E's concern (plan task 15); this task only needs the field to
    /// exist and round-trip.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub replay_from: Option<Value>,
    #[serde(default, skip_serializing_if = "Option::is_none", rename = "_meta")]
    pub meta: Option<Value>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ResumeSessionResponse {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub config_options: Option<Vec<Value>>,
    #[serde(default, skip_serializing_if = "Option::is_none", rename = "_meta")]
    pub meta: Option<Value>,
}
